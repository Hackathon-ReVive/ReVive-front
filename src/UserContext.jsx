import React, { createContext, useState, useEffect } from "react";
import axios from "axios";

export const UserContext = createContext();

export const UserProvider = ({ children }) => {
  const [user, setUser] = useState(null);

  const login = (userData) => {
    setUser(userData);
  };

  const logout = () => {
    setUser(null);
    localStorage.removeItem("token"); // Remove the token from localStorage
  };

  useEffect(() => {
    const checkAuthStatus = async () => {
      try {
        const response = await axios.get("http://localhost:8080/api/auth/me", {
          headers: {
            Authorization: `Bearer ${localStorage.getItem("token")}`,
          },
        });
        setUser(response.data);
      } catch (error) {
        console.error("Not authenticated", error);
      }
    };

    checkAuthStatus();
  }, []);

  return (
    <UserContext.Provider value={{ user, login, logout }}>
      {children}
    </UserContext.Provider>
  );
};
