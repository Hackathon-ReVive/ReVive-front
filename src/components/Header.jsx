import { useContext, useState } from "react";
import { Link } from "react-router-dom";
import { Bars3Icon, XMarkIcon } from "@heroicons/react/24/solid";
import { CartContext } from "../CartContext";
import { UserContext } from "../UserContext";

function Header() {
    const { cart } = useContext(CartContext);
    const { user, logout } = useContext(UserContext);
    const [isOpen, setIsOpen] = useState(false);

    return (
        <header className="bg-white text-emerald-900 font-['Fira_Sans'] flex justify-between items-center p-3">
            <div className="bg-green-100 flex items-center space-x-2 rounded-2xl p-2 m-1">
                <img src="./assets/logo.svg" alt="Logo de ReVive" className="h-10 w-10" />
                <Link to="/home" className="font-bold">ReVive</Link>
            </div>
            <button className="md:hidden block text-emerald-900" onClick={() => setIsOpen(!isOpen)}>
                {isOpen ? <XMarkIcon className="h-8 w-8" /> : <Bars3Icon className="h-8 w-8" />}
            </button>
            <nav className={`flex items-center space-x-2 md:flex ${isOpen ? "flex-col absolute top-16 left-0 w-full bg-white shadow-md" : "hidden md:block"}`}>
                <Link to="/home" className="py-2 px-4">Inicio</Link>
                <Link to="/products" className="py-2 px-4">Productos</Link>
                <Link to="/about" className="py-2 px-4">Sobre nosotres</Link>
                {user ? (
                    <>
                        <span className="py-2 px-4">Bienvenido, {user.username}</span>
                        <button onClick={logout} className="py-2 px-4">Cerrar sesión</button>
                    </>
                ) : (
                    <>
                        <Link to="/register" className="py-2 px-4">Registrarse</Link>
                        <Link to="/login" className="py-2 px-4">Iniciar sesión</Link>
                    </>
                )}
                <Link to="/cart" className="relative">
                    Carrito
                    {cart.length > 0 && (
                        <span className="absolute top-0 right-0 bg-red-500 text-white rounded-full px-2 py-1 text-xs">
                            {cart.length}
                        </span>
                    )}
                </Link>
            </nav>
        </header>
    );
}

export default Header;