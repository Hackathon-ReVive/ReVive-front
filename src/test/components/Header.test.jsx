import "@testing-library/jest-dom";
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import Header from "../../components/Header";
import { CartContext } from "../../CartContext";
import { UserContext } from "../../UserContext";
import { waitFor } from "@testing-library/react";

describe("Header Component", () => {
    test("renders logo and title", () => {
        render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [] }}>
                    <UserContext.Provider value={{ user: null }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );

        expect(screen.getByAltText("Logo de ReVive")).toBeInTheDocument();
        expect(screen.getByText(/ReVive/i)).toBeInTheDocument();
    });

    test("renders navigation links", () => {
        render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [] }}>
                    <UserContext.Provider value={{ user: null }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );

        expect(screen.getByText(/Inicio/i)).toBeInTheDocument();
        expect(screen.getByText(/Productos/i)).toBeInTheDocument();
        expect(screen.getByText(/Sobre nosotres/i)).toBeInTheDocument();
    });

    test("renders login and register buttons when user is not logged in", () => {
        render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [] }}>
                    <UserContext.Provider value={{ user: null }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );

        expect(screen.getByText(/Registrarse/i)).toBeInTheDocument();
        expect(screen.getByText(/Iniciar sesión/i)).toBeInTheDocument();
    });

    test("renders username and logout button when user is logged in", () => {
        const mockLogout = jest.fn();
        render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [] }}>
                    <UserContext.Provider value={{ user: { username: "Lanny" }, logout: mockLogout }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );

        expect(screen.getByText(/Bienvenido, Lanny/i)).toBeInTheDocument();
        expect(screen.getByText(/Cerrar sesión/i)).toBeInTheDocument();
    });

    test("calls logout function when logout button is clicked", () => {
        const mockLogout = jest.fn();
        render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [] }}>
                    <UserContext.Provider value={{ user: { username: "Lanny" }, logout: mockLogout }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );

        fireEvent.click(screen.getByText(/Cerrar sesión/i));
        expect(mockLogout).toHaveBeenCalled();
    });

    test("displays cart count when there are items in the cart", () => {
        render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [1, 2, 3] }}>
                    <UserContext.Provider value={{ user: null }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );

        expect(screen.getByText("3")).toBeInTheDocument();
    });

    test("toggles mobile menu when clicking menu button", async () => {
        render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [] }}>
                    <UserContext.Provider value={{ user: null }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );

        const menuButton = screen.getByRole("button");
        const navMenu = screen.getByRole("navigation");

        fireEvent.click(menuButton);
        expect(navMenu).not.toHaveClass("hidden");

        fireEvent.click(menuButton);

        await waitFor(() => {
            expect(navMenu).toHaveClass("hidden");
        });
    });
    test("matches snapshot", () => {
        const { asFragment } = render(
            <BrowserRouter>
                <CartContext.Provider value={{ cart: [] }}>
                    <UserContext.Provider value={{ user: null }}>
                        <Header />
                    </UserContext.Provider>
                </CartContext.Provider>
            </BrowserRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });
});

