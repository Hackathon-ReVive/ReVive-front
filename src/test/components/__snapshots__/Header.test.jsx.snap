// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Component matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="bg-white text-emerald-900 font-['Fira_Sans'] flex justify-between items-center p-3"
  >
    <div
      class="bg-green-100 flex items-center space-x-2 rounded-2xl p-2 m-1"
    >
      <img
        alt="Logo de ReVive"
        class="h-10 w-10"
        src="./assets/logo.svg"
      />
      <a
        class="font-bold"
        data-discover="true"
        href="/home"
      >
        ReVive
      </a>
    </div>
    <button
      class="md:hidden block text-emerald-900"
    >
      <svg
        aria-hidden="true"
        class="h-8 w-8"
        data-slot="icon"
        fill="currentColor"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          clip-rule="evenodd"
          d="M3 6.75A.75.75 0 0 1 3.75 6h16.5a.75.75 0 0 1 0 1.5H3.75A.75.75 0 0 1 3 6.75ZM3 12a.75.75 0 0 1 .75-.75h16.5a.75.75 0 0 1 0 1.5H3.75A.75.75 0 0 1 3 12Zm0 5.25a.75.75 0 0 1 .75-.75h16.5a.75.75 0 0 1 0 1.5H3.75a.75.75 0 0 1-.75-.75Z"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <nav
      class="flex items-center space-x-2 md:flex hidden md:block"
    >
      <a
        class="py-2 px-4"
        data-discover="true"
        href="/home"
      >
        Inicio
      </a>
      <a
        class="py-2 px-4"
        data-discover="true"
        href="/products"
      >
        Productos
      </a>
      <a
        class="py-2 px-4"
        data-discover="true"
        href="/about"
      >
        Sobre nosotres
      </a>
      <a
        class="py-2 px-4"
        data-discover="true"
        href="/register"
      >
        Registrarse
      </a>
      <a
        class="py-2 px-4"
        data-discover="true"
        href="/login"
      >
        Iniciar sesi√≥n
      </a>
      <a
        class="relative"
        data-discover="true"
        href="/cart"
      >
        Carrito
      </a>
    </nav>
  </header>
</DocumentFragment>
`;
